### ISSUE TITLE
Parsl slows down after hundreds of thousands of jobs 

### ISSUE THREAD

**Issue by reidmcy** [Parsl/parsl#461]
I have a script that requires millions of jobs to be run on parsl and have noticed that after a day or so of running the time per jobs starts increasing. The jobs are independent, nearly identical and randomly ordered so I don't thinks its related to them. I've also changed from `Remote_IPP` to `IPyParallelExecutor` locally to `IPyParallelExecutor` on SLURM and disabled `app_cache`.

Batching jobs has helped reduce the slowdown so it is likely something internal to parsl causing the issue.

My config is:

```
rccNode = Config(
    executors=[
        IPyParallelExecutor(
            label='midway_ipp',
            provider=SlurmProvider(
                'broadwl',
                channel=LocalChannel(),
                #launcher=SrunLauncher(),
                init_blocks=1, # 1
                min_blocks=0, # 0
                max_blocks=10, # 10
                nodes_per_block=1, # 4
                tasks_per_node=28, # 12
                walltime="00:30:00",
                overrides="#SBATCH --mem-per-cpu=2000..."
            ),
        ),
    ],
    app_cache=False,
)
```

**Comment by reidmcy** [comment]
![2018_08_11-23_23_53_count_recent_minutes_cdf](https://user-images.githubusercontent.com/4342916/44007166-7b875fc8-9e56-11e8-9584-9806f4a2bf3b.png)

This shows jobs run vs time, with a restart in the middle and an OLS. You can see that after the restart the line is much steeper, i.e. jobs are running faster.

I think the simplest way to reproduce the issue would be to run one of the test examples in a loop a million times, but I have not tested it.

**Comment by MalcolmUnWang** [comment]
This is interesting. It appears that the run-vs-time line is pretty straight before you restart it. So I assume the running speed drops during the first a hundred thousand and keeps the slower speed before a restart? My first guess is the problem of memory-cache locality. Longer execution results in bigger memory usage, and results in the loss of locality of cache, which causes longer memory access time for the master to pull tasks. Can you show me your code and machine environment you were working on? @reidmcy 

@yadudoc @kylechard @annawoodard If that is a common case, I think a solution might come up with to solve this slow-down. Probably by refreshing where slows us down and not necessary to keep in memory every hundred thousand tasks.

**Comment by reidmcy** [comment]
@MalcolmUnWang the slowdown is nonlinear, it's just not obvious in that picture. Here is a longer time window:
![2018_08_12-06_31_50_count_cdf](https://user-images.githubusercontent.com/4342916/44051745-52686420-9f00-11e8-8a23-82682322976c.png)
 
It seems to transition from linear to nonlinear, presumably as slack in the system runs out and loading/unloading tasks becomes the bottleneck.

**Comment by reidmcy** [comment]
@MalcolmUnWang I can't show you all my code, but the main loops looks something like:
```
    running = {}
    try:
        while not done:
            #Only add maxRunning jobs to the queue at once
            numOldJobs = 0 if oldJobs is None else len(oldJobs)
            while len(running) < config.maxRunning:
                #Load next repo
                batch = []
                names = []
                for i in range(config.reposPerBatch): #Currently loading 5
                    try:
                        #Reads a line from a csv file
                        tDat = next(workingTargets)
                    except StopIteration:
                        #If none left just skip adding
                        doneIter = True
                        break
                    else:
                        #Add repo to list of running jobs
                        tDat['downloadDir'] = currentName
                        batch.append(t)
                        names.append(repoName)
   
            runningNameLists = list(running.keys())
            for names in runningNameLists:
                        if running[names].done():
                            succ, ret_dat, timings = running.pop(names).result()
                            writeResult(repoName, succ, ret_dat)

            #End the loop if all jobs are done and no more can be added
            if doneIter and len(running) < 1:
                done = True
```

**Comment by jmjwozniak** [comment]
Can you watch the memory usage in top and report if it is continually growing?

**Comment by annawoodard** [comment]
The issue isn't just the size in memory, it is that we haven't optimized the search to determine if all of the dependencies are met for a task. I think the entire task tree has to be traversed each time.

**Comment by annawoodard** [comment]
This should have been fixed in #600. We can re-open if we see further issues.
